ironfish/src/wallet/scanner/noteDecryptor.ts

```diff
@@ -0,0 +1,144 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import { BufferMap } from 'buffer-map'
+import { Assert } from '../../assert'
+import { BlockHeader } from '../../primitives'
+import { Transaction } from '../../primitives/transaction'
+import { WorkerPool } from '../../workerPool'
+import { DecryptedNote, DecryptNotesOptions } from '../../workerPool/tasks/decryptNotes'
+import { Account } from '../account/account'
+import { Queue } from '../scanner/queue'
+
+// account id -> transaction hash -> decrypted note
+export type AccountTransactionNotesMap = Map<string, BufferMap<Array<DecryptedNote>>>
+
+export type DecryptNotesFromTransactionsCallback = (
+  accountTransactionNotesMap: AccountTransactionNotesMap,
+) => Promise<void>
+
+export class BackgroundNoteDecryptor {
+  private isStarted = false
+
+  private triggerFlushed: (() => void) | null = null
+  private triggerStopped: (() => void) | null = null
+
+  private onFlushed: Promise<void> = Promise.resolve()
+  private onStopped: Promise<void> = Promise.resolve()
+
+  private readonly workerPool: WorkerPool
+  private readonly options: DecryptNotesOptions
+  private readonly decryptQueue: Queue<{
+    promise: Promise<ReadonlyMap<string, ReadonlyArray<DecryptedNote | null>>>
+    transactions: ReadonlyArray<Transaction>
+    callback: DecryptNotesFromTransactionsCallback
+  }>
+
+  constructor(workerPool: WorkerPool, options: DecryptNotesOptions) {
+    this.workerPool = workerPool
+    this.options = options
+    this.decryptQueue = new Queue(Math.max(32 * workerPool.numWorkers, 1))
+  }
+
+  start() {
+    if (!this.isStarted) {
+      this.isStarted = true
+      this.onStopped = new Promise((resolve) => (this.triggerStopped = resolve))
+      void this.decryptLoop()
+    }
+  }
+
+  stop() {
+    this.isStarted = false
+    if (this.triggerStopped) {
+      this.triggerStopped()
+    }
+  }
```

Comment:
We did talk about this offline but I want to repost my final thoughts her as well. After looking at the code, I think this code works as expected. The primary issue I have is that this code doesn't manage it's promises very well causing the node to hault up when you're doing things like... telling the node to stop, and aborting a scan.

I can see why. If the abort controller for a scan is triggered, it waits for the decryptor to flush and then it stops it. I think this is an issue with the current version but I would definitely want it fixed in a future PR.

Things that have queue's of async work in general should manage those items, and cancel them appropriately when the system is told to stop. Abort's should be propagated through the sub-modules and ideally cancel async operations if they are abled to and fully wait on them to be aborted so no hanging promises are left which will keep a NodeJS process alive after the main loop has exited.